examples/TestEm3/electrons.cu

```diff
@@ -247,18 +258,86 @@ static __device__ __forceinline__ void TransportElectrons(Track *electrons, cons
-    // ready to be used.
-    newRNG.Advance();
     // Also advance the current RNG state to provide a fresh round of random
     // numbers after MSC used up a fair share for sampling the displacement.
     currentTrack.rngState.Advance();
 
+    if constexpr (IsElectron) g_nextInteractionForEl[i] = winnerProcessIndex;
+    else g_nextInteractionForPos[i] = winnerProcessIndex;
+    assert(i < (IsElectron ?
+        sizeof(g_nextInteractionForEl)/sizeof(g_nextInteractionForEl[0])
+      : sizeof(g_nextInteractionForPos)/sizeof(g_nextInteractionForPos[0]) ));
+  }
+}
+
+
+// Instantiate kernels for electrons and positrons.
+__global__ __launch_bounds__(ThreadsPerBlock, MinBlocksPerSM)
+void TransportElectrons(Track *electrons, const adept::MParray *active, Secondaries secondaries,
+                        adept::MParray *activeQueue, GlobalScoring *globalScoring,
+                        ScoringPerVolume *scoringPerVolume)
+{
+  TransportElectrons</*IsElectron*/ true>(electrons, active, secondaries, activeQueue, globalScoring, scoringPerVolume);
+}
+__global__ __launch_bounds__(ThreadsPerBlock, MinBlocksPerSM)
+void TransportPositrons(Track *positrons, const adept::MParray *active, Secondaries secondaries,
+                        adept::MParray *activeQueue, GlobalScoring *globalScoring,
+                        ScoringPerVolume *scoringPerVolume)
+{
+  TransportElectrons</*IsElectron*/ false>(positrons, active, secondaries, activeQueue, globalScoring,
+                                           scoringPerVolume);
+}
+
+template<bool IsElectron, int ProcessIndex>
+__global__
+void ComputeInteraction(Track *electrons, const adept::MParray *active, Secondaries secondaries,
+                        adept::MParray *activeQueue, GlobalScoring *globalScoring,
+                        ScoringPerVolume *scoringPerVolume)
+{
+  constexpr unsigned int sharedSize = 12250;
+  __shared__ int candidates[sharedSize];
+  __shared__ unsigned int counter;
+  __shared__ unsigned int noopCounter;
+  counter = 0;
+  noopCounter = 0;
+
+  for (int i = threadIdx.x; i < sharedSize; i += blockDim.x) {
+    candidates[i] = -1;
+  }
+  __syncthreads();
+
+  const int activeSize = active->size();
+  assert(activeSize < sizeof(g_nextInteractionForEl)/sizeof(g_nextInteractionForEl[0]));
+  for (int i = blockIdx.x * blockDim.x + threadIdx.x; i < activeSize; i += blockDim.x * gridDim.x) {
+    const auto winnerProcess = IsElectron ? g_nextInteractionForEl[i] : g_nextInteractionForPos[i];
+
+    if (winnerProcess == ProcessIndex) {
+      const auto destination = atomicInc(&counter, (unsigned int)-1);
+      candidates[destination % sharedSize] = i;
+    } else atomicInc(&noopCounter, (unsigned int)-1);
+  }
+
+  __syncthreads();
+  assert(counter < sharedSize);
+  if (threadIdx.x == 0 && counter >= sharedSize) {
+    printf("Error: Shared queue for %d exhausted to %d in %s:%d\n", ProcessIndex, counter, __FILE__, __LINE__);
+    asm("trap;");
+  }
```

Comment:
Could you not just wrap another loop around the entire kernel that fills a chunk of shared memory, processes it, and then starts with the next chunk?